torch/csrc/autograd/python_engine.cpp

```diff
@@ -392,8 +392,25 @@ PyObject* THPEngine_queue_callback(PyObject* self, PyObject* _callback) {
   engine.queue_callback([callback]() {
     pybind11::gil_scoped_acquire gil;
     THPObjectPtr result{PyObject_CallFunctionObjArgs(callback.get(), nullptr)};
-    if (!result)
-      throw python_error();
+    if (!result) {
+      // Note [ Persisting PyErr state across autograd engine threads ]
+      //
+      // Since the autograd engine is multi-threaded, and Python error state is
+      // local to each thread, it must preserve the python error from the worker
+      // thread and rethrow it as-is in the calling thread. This is done via
+      // persisting the error in the two places that can encounter Python
+      // errors: (1) evaluate function and (2) queued callbacks.
+      //
+      // TODO: the engine is not actually responsible for persisting the error
```

Comment:
So all of these can be removed?

Reply:
Errr I should this comment more precise. The engine is not responsible for persisting the error particularly in the custom Function and backward hooks cases today, but I haven't looked at any other cases (torch dispatch, PyInterpreter cases) so if we want to remove the persist in the engine, those cases also need to be checked that they aren't being regressed. 

I'm not sure its entirely worth it to look into more of these cases if we already believe that the best solution is for persist to be handled in the engine. Like if someone wants to fix the DistEngine torch dispatch error propagation for example (I don't know if its actually broken), I'd prefer that they fix the issue systematically rather than have to exhaustively insert the persist at every individual call site.